sale_order_blanket_order/models/sale_order_line.py

```diff
@@ -0,0 +1,554 @@
+        - The product must be part of the linked blanket order.
+        - The quantity to procure must be less than or equal to the quantity
+          remaining to deliver in the linked blanket order for this product.
+        """
+        matching_dict = self._get_blanket_lines_for_call_off_lines_dict()
+        for call_of_lines, blanket_lines in matching_dict.values():
+            if not blanket_lines:
+                line = call_of_lines[0]
+                raise ValidationError(
+                    _(
+                        "The product is not part of linked blanket order. "
+                        "(Product: '%(product)s', Order: '%(order)s', "
+                        "Blanket Order: '%(blanket_order)s')",
+                        product=line.product_id.display_name,
+                        order=line.order_id.name,
+                        blanket_order=line.blanket_order_id.name,
+                    )
+                )
+
+            qty_remaining_to_procure = sum(
+                blanket_lines.mapped("call_off_remaining_qty")
+            )
+            qty_to_procure = sum(call_of_lines.mapped("qty_to_deliver"))
+            if (
+                float_compare(
+                    qty_to_procure,
+                    qty_remaining_to_procure,
+                    precision_rounding=call_of_lines[0].product_uom.rounding,
+                )
+                > 0
+            ):
+                raise ValidationError(
+                    _(
+                        "The quantity to procure is greater than the quantity "
+                        "remaining to deliver in the linked blanket order for "
+                        "this product. (Product: '%(product)s', Order: "
+                        "'%(order)s', Blanket Order: '%(blanket_order)s')",
+                        product=call_of_lines[0].product_id.display_name,
+                        order=call_of_lines[0].order_id.name,
+                        blanket_order=call_of_lines[0].blanket_order_id.name,
+                    )
+                )
+
+    @api.constrains("order_type", "price_unit")
+    def _check_call_off_order_line_price(self):
+        price_precision = self.env["decimal.precision"].precision_get("Product Price")
+        for line in self:
+            if line.order_type == "call_off" and not float_is_zero(
+                line.price_unit, precision_digits=price_precision
+            ):
+                raise ValidationError(
+                    _(
+                        "The price of a call-off order line must be 0.0. "
+                        "(Order: '%(order)s', Product: '%(product)s')",
+                        order=line.order_id.name,
+                        product=line.product_id.display_name,
+                    )
+                )
+
+    @api.constrains(
+        "blanket_validity_start_date",
+        "blanket_validity_end_date",
+        "product_id",
+    )
+    def _check_blanket_product_not_overlapping(self):
```

Comment:
Otherwise it becomes difficult to know which blanket line to choose in case of automatic creation of call off order at normal order confirmation. (Reading your comment and the list of declared tiggers , I see a potential bug since all the fields used to detect 'double' are not into the list of triggers )

Reply:
don